ambry-store/src/main/java/com.github.ambry.store/Offset.java

```diff
@@ -0,0 +1,127 @@
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+
+/**
+ * Denotes an offset inside the log.
+ */
+class Offset implements Comparable<Offset> {
+  private final String name;
+  private final long offset;
+
+  private static final short CURRENT_VERSION = 0;
+  private static final int VERSION_SIZE = 2;
+  private static final int NAME_LENGTH_SIZE = 4;
+  private static final int OFFSET_SIZE = 8;
+
+  /**
+   * Construct an Offset that refers to a position in Log.
+   * @param name the name of segment being referred to.
+   * @param offset the offset within the segment.
+   * @throws IllegalArgumentException if {@code name} is {@code null} or empty or {@code offset} < 0.
+   */
+  Offset(String name, long offset) {
+    if (name == null || name.isEmpty() || offset < 0) {
+      throw new IllegalArgumentException("Name [" + name + "] is null/empty or offset [" + offset + "] < 0");
+    }
+    this.name = name;
+    this.offset = offset;
+  }
+
+  /**
+   * Constructs an Offset from a stream.
+   * @param stream the {@link DataInputStream} that will contain the serialized form of this object.
+   * @throws IllegalArgumentException if {@code name} is {@code null} or empty or {@code offset} < 0.
+   * @throws IOException if there are I/O problems reading from the stream.
+   */
+  static Offset fromBytes(DataInputStream stream)
+      throws IOException {
+    String name;
+    long offset;
+    int version = stream.readShort();
+    switch (version) {
+      case 0:
+        name = Utils.readIntString(stream, StandardCharsets.UTF_8);
+        offset = stream.readLong();
+        break;
+      default:
+        throw new IllegalArgumentException("Unrecognized version [" + version + "] of Offset");
+    }
+    return new Offset(name, offset);
+  }
+
+  /**
+   * @return the name of the log segment for which the offset provided by {@link #getOffset()} is valid.
+   */
+  String getName() {
+    return name;
+  }
+
+  /**
+   * @return the offset in the log segment with name provided by {@link #getName()}.
+   */
+  long getOffset() {
+    return offset;
+  }
+
+  /**
+   * Converts this object into its byte representation.
+   * @return the byte representation of this object.
+   */
+  byte[] toBytes() {
+    byte[] nameBytes = name.getBytes(StandardCharsets.UTF_8);
+    int size = VERSION_SIZE + NAME_LENGTH_SIZE + nameBytes.length + OFFSET_SIZE;
+    byte[] bytes = new byte[size];
+    ByteBuffer buf = ByteBuffer.wrap(bytes);
+    buf.putShort(CURRENT_VERSION);
+    buf.putInt(nameBytes.length);
+    buf.put(nameBytes);
+    buf.putLong(offset);
+    return bytes;
+  }
+
+  @Override
+  public int compareTo(Offset o) {
+    // TODO (Log Segmentation): once LogSegment is coded and can resolve the relative position of a log segment based
```

Comment:
Good points.

**How will this class be used?**: It will refer to an offset in a log that in turn contains multiple log segments. Since the log segments themselves are individual files, an offset that is just a `long` will not be of use. The `Offset` class therefore is a combination of the name of the log segment and an offset and together they can find a particular offset in the abstracted log.

**Why is this a TODO?**: It is not as simple as a lexicographic sort since the log segment name has the form "position_genno_log". There are also some subtleties around handling the generation numbers. I don't want to code that in multiple places but instead let the log segment implementation resolve this. 

**Why is external to LogSegment?**: Because the `LogSegment` itself doesn't care about this offset. It is of importance to the components using the abstracted log.

**Why is this being introduced now?**:
There are 3-4 broad stages for segmentation:
1. Abstracting offsets (this PR).
2. Using abstracted offsets at places where the abstracted log will be used (`Journal`, indexes, `BlobStore` etc) (next PR).
3. Segmenting the log.
4. Adding support for this in the index. (might be merged with 3).

You are right that this PR alone might lack context but merging it with any of the other stages creates a lot of changes and moving parts that will be hard to review. These stages obviously have a lot of interdependencies so you will see a lot of "TODOs" for the next stages. So this is done in the interest of quick and easier reviewing where each of these components may be reviewed in isolation. 

This goes back to the question you had about having a design/implementation document (which I am working on) and that will provide the necessary context. In the case of this particular PR, I thought that change was isolated enough to not require the document.

There is another way to achieve this implementation and that is to code the `LogSegment` and all related changes under the assumption that there is only one segment and then change the code so that that assumption is removed. I thought that would be the harder way.